src/ariel-os-nrf/src/uart.rs

```diff
@@ -0,0 +1,376 @@
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum DataBits {
+    /// 8 bits per character.
+    Data8,
+}
+
+impl From<ariel_os_embassy_common::uart::DataBits<Self>> for DataBits {
+    fn from(databits: ariel_os_embassy_common::uart::DataBits<Self>) -> DataBits {
+        match databits {
+            ariel_os_embassy_common::uart::DataBits::Hal(bits) => bits,
+            ariel_os_embassy_common::uart::DataBits::Data8 => DataBits::Data8,
+        }
+    }
+}
+/// UART number of stop bits.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum StopBits {
+    /// One stop bit.
+    Stop1,
+}
+
+impl From<ariel_os_embassy_common::uart::StopBits<Self>> for StopBits {
+    fn from(stopbits: ariel_os_embassy_common::uart::StopBits<Self>) -> Self {
+        match stopbits {
+            ariel_os_embassy_common::uart::StopBits::Hal(stopbits) => stopbits,
+            ariel_os_embassy_common::uart::StopBits::Stop1 => StopBits::Stop1,
+        }
+    }
+}
+
+/// Parity bit.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum Parity {
+    /// No parity bit.
+    None,
+    /// Even parity bit.
+    Even,
+}
+
+fn from_parity(parity: Parity) -> embassy_nrf::uarte::Parity {
+    match parity {
+        Parity::None => embassy_nrf::uarte::Parity::EXCLUDED,
+        Parity::Even => embassy_nrf::uarte::Parity::INCLUDED,
+    }
+}
+
+impl From<ariel_os_embassy_common::uart::Parity<Self>> for Parity {
+    fn from(parity: ariel_os_embassy_common::uart::Parity<Self>) -> Self {
+        match parity {
+            ariel_os_embassy_common::uart::Parity::Hal(parity) => parity,
+            ariel_os_embassy_common::uart::Parity::None => Self::None,
+            ariel_os_embassy_common::uart::Parity::Even => Self::Even,
+        }
+    }
+}
+
+macro_rules! define_uart_drivers {
+    ($( $interrupt:ident => $peripheral:ident + $timer:ident + $ppi_ch1:ident + $ppi_ch2:ident + $ppi_group:ident),* $(,)?) => {
+        $(
+            /// Peripheral-specific UART driver.
+            pub struct $peripheral<'d> {
+                uart: BufferedUarte<'d, peripherals::$peripheral, peripherals::$timer>,
+            }
+
+            // Make this struct a compile-time-enforced singleton: having multiple statics
+            // defined with the same name would result in a compile-time error.
+            paste::paste! {
+                #[allow(dead_code)]
+                static [<PREVENT_MULTIPLE_ $peripheral>]: () = ();
+                #[allow(dead_code)]
+                static [<PREVENT_MULTIPLE_ $timer>]: () = ();
+                #[allow(dead_code)]
+                static [<PREVENT_MULTIPLE_ $ppi_ch1>]: () = ();
+                #[allow(dead_code)]
+                static [<PREVENT_MULTIPLE_ $ppi_ch2>]: () = ();
+                #[allow(dead_code)]
+                static [<PREVENT_MULTIPLE_ $ppi_group>]: () = ();
```

Comment:
Only the first singleton here is needed to prevent multiple instances. The others prevent accidental configuration issues, e.g. using `PPI_CH18` twice across the UART instances.